controller/resource.go

```diff
@@ -174,17 +176,66 @@ func (t *Task) Upsert() error {
 	}
 
 	// Otherwise, create one.
-	t.Log.Info("creating service token", "id", serviceName)
 	err = t.updateServiceToken(serviceName, secretName)
 	if err != nil {
 		return fmt.Errorf("updating service token: %w", err)
 	}
-	t.Log.Info("service token created successfully", "id", serviceName)
 
 	return nil
 }
 
 func (t *Task) Delete() error {
+	if !t.isMeshTask() {
+		t.Log.Info("skipping non-mesh task", "id", *t.Task.TaskDefinitionArn)
+		return nil
+	}
+
+	serviceName, err := t.parseFamilyNameFromTaskDefinitionARN()
+	if err != nil {
+		return fmt.Errorf("parsing service name: %w", err)
+	}
+
+	services, _, err := t.ConsulClient.Catalog().Service(serviceName, "", nil)
+	if err != nil {
+		return fmt.Errorf("getting service from Consul: %w", err)
+	}
+	if len(services) != 0 {
+		t.Log.Info("skipping because service still exists", "id", *t.Task.TaskDefinitionArn)
+		return nil
```

Comment:
👍 

Only thought: We return `nil` here, which causes this TaskDefinitionARN to be removed from `c.resourceState`. So maybe there's risk the controller forgets to clean up a token... It would be a narrow window, for example:

* The last task of a task family is stopped
* consul-client begins to shutdown
* `reconcile` runs and does not find the stopped task, so calls Delete
* Delete sees the service in Consul, and skips the token cleanup
* The TaskDefinitionARN is removed from c.resourceState
* consul-client removes the service from Consul, and then leaves the cluster

Maybe worth a comment in the code, but it seems rare. And it wouldn't matter if you later start more tasks for the same family, since they would reuse the token.

So I am good with this!

Reply:
> Only thought: We return nil here, which causes this TaskDefinitionARN to be removed from c.resourceState. So maybe there's risk the controller forgets to clean up a token.

Yeah, this is intentional. I wasn't able to create a race condition that you're describing, but if there is one, then it'd be safer to keep the token. I'll add a comment though!